lib/ansible/template/__init__.py

```diff
@@ -152,10 +157,186 @@ def _count_newlines_from_end(in_str):
         while in_str[j] == '\n':
             j -= 1
         return i - 1 - j
+    except TypeError:
+        return 0
     except IndexError:
         # Uncommon cases: zero length string and string containing only newlines
         return i
 
+
+def ansible_j2_concat(invals):
+    '''
+    Jinja's native concat is just a list join. The list comes from result of
+    templating all the nodes. If the return type is expected to be
+    native, the list is artificial and we should return just the first
+    element.
+    '''
+    invals = [x for x in invals]
+    if isinstance(invals, list):
+        if len(invals) == 1:
+            # break out the single value
+            invals = invals[0]
+        elif len(invals) > 1:
+            # cast to unicode and join
+            invals = u''.join([u'%s' % x for x in invals])
+    return invals
```

Comment:
Okay, jctanner says that this is always a generator.  So this is a better implementation:

``` python
from itertools import chain
start = []
for val in invals:
    start.append(val)
    if len(start) > 1:
        break
else:
    return start[0]
return u''.join(chain(start, invals))
```